### PR TITLE
Make `Sift#filter_errors` return `ActiveModel::Errors` instead of `Hash`

### PR DESCRIPTION
**Pull request by yevhenii-ponomarenko**
# Summary

This PR achieves several things:
- Updates gem codebase to require `rails >= 6.1`. (This allowed tests to run on Rails 7.0)
- Remove support for Rails 4.
- Explicitly states `net-http` as a dependency. This gem is default in Ruby 3. Adding it removed several warning messages.
- BREAKING CHANGE: `Sift#filter_errors` now returns an instance of `ActiveModel::Errors` instead of `Hash`.

This solidifies Rails as dependency of Sift.

# Testing
- [x] `bundle exec rake test` all green.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 ## Unreleased
 
+## 0.17.0
```
Comment (kkohrt) [inline CHANGELOG.md:3]: Let's go to 1.0 with this change

**Review by dervish86 — APPROVED** [review]

**Comment by joe-sharp** [comment]
@lessthanjacob and/or @grahamleslie would you mind taking a look at this PR?

**Comment by lessthanjacob** [comment]
Apologies for the delay here. 

Could you add some detail on why we'd want to start returning an instance `ActiveModel::Error` instead of a hash here? Based on how this has been implemented (we're essentially defining a dummy `ActiveRecord` model to attach filter validations to), I'd consider the fact that we're leveraging `ActiveModel::Validations` an implementation detail that doesn't need to be exposed in this gem's interface.

Additionally, the lack of required test changes here has me concerned that we're missing some coverage, given that this should be a breaking change.

**Comment by lessthanjacob** [comment]
@yevhenii-ponomarenko Thanks for the background info!

Long term, I'd like to see us move away from the dependency on `ActiveModel::Error` here, and instead encapsulate that detail behind a bespoke `Sift::FilterValidationError` (or similar object) if possible. This gem _is_ intentionally heavily dependent on `ActiveRecord` as a way to provide a simple integration with Rails controllers and the models they delegate to, but I think we could improve the extensibility and maintainability of this code by not coupling the interface to its implementation. 

However, if this is currently a blocker on our upgrade path, I'm in support of this change, especially given the existing coupling we'd have to unwind to change the interface. If you could add some additional test coverage, that would really be appreciated!

File: lib/procore-sift.rb [hunk 2]
```diff
@@ -32,7 +32,7 @@ def filters_valid?
   end
 
   def filter_errors
-    filter_validator.errors.messages
```
Comment (kkohrt) [inline lib/procore-sift.rb:35]: The one drawback is that we've broken the contract published in the [README on the homepage](https://github.com/procore/sift#developer-usage)
```
def render_filter_errors
  render json: { errors: filter_errors }, status: :bad_request && return
end
```

The README should be updated to show the correct syntax for the 1.x version of the code (once determined), as well as the legacy syntax--since people with questions often come and give a quick look at the README (or even have it linked in some other doc) so it should be able to inform them of both the current and past valid syntax.

However, in reflecting on #lessthanjacob's comments, as well as the [other statement in the README](https://github.com/procore/sift#without-rails) that suggests we are trying to move away from a Rails dependency, I think the better course of action would be to leave this interface as-is, which would mean adding code to gracefully transform Rails 7 errors into the expected hash format .
Then, for accessing the raw errors--assuming that is one of the objectives--create a new public accessor that provides that. 
`def raw_filter_errors` or `def active_model_filter_errors` 
or something like that.
In this way we maintain current behavior while allowing for new access to the underlaying error objects.
And, in version 2.0, when maybe we drop Rails dependancies, we can consider dropping the raw object access (or maybe making it a wrapper object that quacks like an ActiveModel::Error object).

**Review by kkohrt — COMMENTED** [review]
TL;DR: I'd propose adding an new endpoint rather than breaking the existing endpoint

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,11 @@
 ## Unreleased
 
+## 1.0.0
+
+- Breaking change: Bump `rails` dependency to `>=6.1`
+- Breaking change: Make `Sift#filter_errors` return `ActiveModel::Errors` object instead of `Hash`
```
Comment (kkohrt) [inline CHANGELOG.md:6]: This is not longer breaking. Rails 7 will return a hash

**Review by kkohrt — COMMENTED** [review]
I greatly appreciate the effort here. Especially the thorough documentation and testing. But this kind of sums up the final issue:
>Though we have some plans to remove Rails as a dependency, currently you can access raw `ActiveModel::Errors` object by using `Sift#active_model_filter_errors`.
We plan to remove this for Sift 2.0 release.

Even though I originally suggested it, I think a public accessor to ActiveModel::Errors would block the original goal of making the gem Rails independent. 

Having said that, many of the other changes actually move us towards that goal by removing `Rails.version` from the code.
If you want to support that effort further, you could drop 
`s.add_dependency "rails", ">= 6.1"`
And reduce the runtime dependencies to just
* activemodel (used for validation)
* activerecord (used only to coerce booleans--which could be replaced with some effort)
* activesupport (used only for ActiveSupport::Concern--which could be replaced easily)

We can then reduce Rails to a development dependency only.
